packages/contracts/contracts/gatekeepers/AnonAadhaarGatekeeper.sol

```diff
@@ -0,0 +1,93 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.20;
+
+import { Ownable } from "@openzeppelin/contracts/access/Ownable.sol";
+
+import { SignUpGatekeeper } from "./SignUpGatekeeper.sol";
+import { IAnonAadhaar } from "../interfaces/IAnonAadhaar.sol";
+
+/// @title AnonAadhaarGatekeeper
+/// @notice A gatekeeper contract which allows users to sign up to MACI
+/// only if they can prove they are valid Aadhaar owners.
+/// @dev Please note that once a identity is used to register, it cannot be used again.
+/// This is because we store the nullifier of the proof.
+contract AnonAadhaarGatekeeper is SignUpGatekeeper, Ownable(msg.sender) {
+  /// @notice The anonAadhaar contract
+  IAnonAadhaar public immutable anonAadhaarContract;
+
+  /// @notice The address of the MACI contract
+  address public maci;
+
+  /// @notice The time window in which the proof is valid
+  uint256 public immutable proofValidTime;
+
+  /// @notice The registered identities
+  mapping(uint256 => bool) public registeredAadhaars;
+
+  /// @notice Errors
+  error ZeroAddress();
+  error OnlyMACI();
+  error AlreadyRegistered();
+  error InvalidProof();
+  error proofTooOld(uint timestamp);
+
+  /// @notice Create a new instance of the gatekeeper
+  /// @param _anonAadhaarVerifierAddr The address of the anonAadhaar contract
+  constructor(address _anonAadhaarVerifierAddr, uint256 _proofValidTime) payable {
+    if (_anonAadhaarVerifierAddr == address(0)) revert ZeroAddress();
+    anonAadhaarContract = IAnonAadhaar(_anonAadhaarVerifierAddr);
+    proofValidTime = _proofValidTime;
+  }
+
+  /// @notice Adds an uninitialised MACI instance to allow for token signups
+  /// @param _maci The MACI contract interface to be stored
+  function setMaciInstance(address _maci) public override onlyOwner {
+    if (_maci == address(0)) revert ZeroAddress();
+    maci = _maci;
+  }
+
+  /// @notice Register an user if they can prove anonAadhaar proof
+  /// @dev Throw if the proof is not valid or just complete silently
+  /// @param _data The ABI-encoded data containing nullifierSeed, nullifier, timestamp, signal, revealArray, and groth16Proof.
+  function register(address /*_user*/, bytes memory _data) public override {
```

Comment:
I see one problem is that anyone can just front run this transaction and signup with someone else adhaar. Could we add the user address as signal and check that?